Implement pure, trusted, snapshot_equality and field access

tillarnold opened the pull request:
Builds on top of #13. To see what changed since #13 see https://github.com/tillarnold/prusti-dev/compare/till_rebuilt_1..tillarnold:prusti-dev:till_rebuilt_2

- Allow methods and pure functions to call pure functions
- Implement `#[trusted]`, simply by not encoding bodies for functions with the trusted attribute.
- do not create a method version of a function if it is both `#[trusted]` and `#[pure]`
- Implement field access in pure functions.
- Implement snapshot equality.
- Changes some of the encoding in `spec.rs`. the returned value in the MIR version of post conditions appears to be the last argument and not the first. So I changed that in `spec.rs`. Otherwise the types wouldn't match up.

(This replaces #5 #6 and #10)


([This](https://github.com/tillarnold/prusti-dev/compare/4f70d403edf9feecb7df8a97bb704f5aa14d9893..8e28cbfb6fbd3f9ee2e033989d8e3077bee71cce) is the diff between this branch and the version of #13 before the other changes were merged in. This link is here for me to compare and see what needs to be done to allow merging again)

zgrannan committed:
Introduce functionidentifier struct

zgrannan committed:
Fix bugs

zgrannan committed:
WIP

zgrannan committed:
WIP

zgrannan committed:
Cleanup

zgrannan committed:
refactor

zgrannan committed:
refactor

zgrannan committed:
Use type encoder in more places

zgrannan committed:
cleanup

zgrannan committed:
cleanup

JonasAlaif committed:
Merge branch 'rewrite-2023' into zgrannan/functionidentifier

JonasAlaif committed:
Resolve merge issues

JonasAlaif committed:
Clean up the type encoder

tillarnold committed:
Miscellaneous small fixes

tillarnold committed:
Implement pure, trusted, snapshot_equality and field access

tillarnold committed:
Cleanup

zgrannan committed:
Update FunctionIdentifier

zgrannan committed:
Refactor to not use reify

zgrannan committed:
remove unused imports

JonasAlaif committed:
Add Predicate and Methods and clean things up

JonasAlaif committed:
Bugfix

JonasAlaif committed:
Merge branch 'zgrannan/functionidentifier' into till_rebuilt_1

JonasAlaif committed:
Merge fixes

JonasAlaif committed:
Builtin encoder improvements

JonasAlaif committed:
Merge branch 'rewrite-2023' into till_rebuilt_1

JonasAlaif committed:
Fix merge

JonasAlaif committed:
Bugfix

JonasAlaif committed:
Prevent use of `mk_func_app`

JonasAlaif committed:
Clean up `TerminatorKind::Assert`

JonasAlaif committed:
Fix review issue

JonasAlaif committed:
Various improvements

JonasAlaif commented:
I am currently merging in the updates from https://github.com/tillarnold/prusti-dev/tree/till_rebuilt_1 so do not fix the issues I raise, they are just for your fyi to avoid repeating them in the future

JonasAlaif commented on prusti-encoder/src/encoders/mir_impure.rs:
```diff
@@ -161,29 +161,49 @@ impl TaskEncoder for MirImpureEncoder {
+                 .any(|item| {
+                     item.path.segments.len() == 2
+                         && item.path.segments[0].ident.as_str() == "prusti"
+                         && item.path.segments[1].ident.as_str() == "trusted"
```
There is an established system for checking if a function is trusted, please do not circumvent it. It's the same thing you used to find if a function is `pure` here: https://github.com/tillarnold/prusti-dev/blob/fa4aa23074ba6712abfe8bc7f0715fc509be5bc0/prusti-encoder/src/lib.rs#L145-L156
This is bad because if we were to change the text of the "trusted" tag then this would break.

JonasAlaif commented on prusti-encoder/src/encoders/mir_impure.rs:
```diff
@@ -802,22 +823,87 @@ impl<'vir, 'enc> mir::visit::Visitor<'vir> for EncoderVisitor<'vir, 'enc> {
+                    .map(|attr| attr.get_normal_item()).any(|item| 
+                        item.path.segments.len() == 2
+                        && item.path.segments[0].ident.as_str() == "prusti"
+                        && item.path.segments[1].ident.as_str() == "pure"
```
Same as other comment on "trusted"

JonasAlaif commented on prusti-encoder/src/encoders/mir_impure.rs:
```diff
@@ -802,22 +823,87 @@ impl<'vir, 'enc> mir::visit::Visitor<'vir> for EncoderVisitor<'vir, 'enc> {
+                        let return_type = self.deps
+                            .require_ref::<crate::encoders::TypeEncoder>(body.return_ty())
+                            .unwrap();
+                        return_type.method_assign
```
It would be better to just extend `MirFunctionEncoder` to also return the `return` `vir::Type` since it will already have it available instead of going through the convoluted `self.vcx.body.borrow_mut().get_impure_fn_body_identity`.

JonasAlaif commented on prusti-encoder/src/encoders/mir_impure.rs:
```diff
@@ -802,22 +823,87 @@ impl<'vir, 'enc> mir::visit::Visitor<'vir> for EncoderVisitor<'vir, 'enc> {
+                };
+
+                let enc = self.encode_operand_snap(cond);
+                let enc = self.vcx.mk_func_app("s_Bool_val", &[enc]);
```
I will be disabling the use of `mk_func_app` completely, you can replace a line like this with e.g. https://github.com/Aurel300/prusti-dev/blob/2ac6b644c7bb61763bcfb7f7671109e5eedaae49/prusti-encoder/src/encoders/mir_pure.rs#L549-L551 in the future

JonasAlaif commented on prusti-encoder/src/encoders/mir_pure.rs:
```diff
@@ -468,6 +469,66 @@ impl<'vir, 'enc> Encoder<'vir, 'enc>
+                        let is_pure = normal_attrs.iter().any(|item| 
+                            item.path.segments.len() == 2
+                            && item.path.segments[0].ident.as_str() == "prusti"
+                            && item.path.segments[1].ident.as_str() == "pure"
```
Same comment as for "trusted"

JonasAlaif commented on prusti-encoder/src/encoders/mir_pure.rs:
```diff
@@ -734,31 +824,55 @@ impl<'vir, 'enc> Encoder<'vir, 'enc>
+        let mut expr = local;
+
+        for elem in place.projection {
+            (partent_ty, expr) = self.encode_place_element(partent_ty, elem, expr);
```
Better to use the `fold` function of `Iterator` here

JonasAlaif commented on prusti-encoder/src/encoders/pure/spec.rs:
```diff
@@ -105,14 +108,14 @@ impl TaskEncoder for MirSpecEncoder {
                     &[expr],
                 );
-                expr.reify(vcx, (*spec_def_id, &pre_args[1..]))
+                expr.reify(vcx, (*spec_def_id, &pre_args.split_last().unwrap().1))
```
Indexing with `pre_args[..pre_args.len() - 1]` is clearer

JonasAlaif committed:
Merge branch 'till_rebuilt_1' into till_rebuilt_2

JonasAlaif committed:
Resolve merge issues

JonasAlaif committed:
Fix join point algorithm

JonasAlaif committed:
Switch to `IndexVec`

JonasAlaif committed:
Merge branch 'till_rebuilt_1' into till_rebuilt_2

JonasAlaif committed:
Fix issues in `mir_builtin` and some lifetimes

- The `mir_builtin` should now be complete and not need to be changed
- Start work on separating `'tcx` and `'vir`

JonasAlaif committed:
Resolve review comments

JonasAlaif committed:
Merge branch 'till_rebuilt_1' into till_rebuilt_2

JonasAlaif committed:
Merge branch 'rewrite-2023' into till_rebuilt_2

JonasAlaif committed:
Fix merge issues

JonasAlaif committed:
Cleanup

JonasAlaif committed:
Some more cleanup

JonasAlaif committed:
Cleanup pure place encoding

JonasAlaif commented:
@Aurel300 ready to merge after you review

Aurel300 commented on prusti-encoder/src/encoders/spec.rs:
```diff
@@ -35,6 +35,16 @@ where
+{
+    DEF_SPEC_MAP.with_borrow(|def_spec: &Option<DefSpecificationMap>| {
+        let def_spec = def_spec.as_ref().unwrap();
+        def_spec.get_proc_spec(&def_id).map(|spec| &spec.base_spec).map(f)
```
Let's at least add a `TODO` here: `base_spec` is not what we always want.

Aurel300 commented on prusti-encoder/src/encoders/mir_impure.rs:
```diff
@@ -64,7 +68,7 @@ impl TaskEncoder for MirImpureEncoder {
     }
 
-    fn do_encode_full<'tcx: 'vir, 'vir>(
+    fn do_encode_full<'tcx: 'vir, 'vir: 'tcx>(
```
What forced the invariance?

JonasAlaif replied:
I hadn't been able to figure it out, so that's why it was like this. Now I've tracked it down to having `'vir` here: https://github.com/Aurel300/prusti-dev/pull/14/files#diff-7e0bdbbcc42baa44b6008e26f05fcd4e65edf5b564fe207827b8dc8850d542f8R117

Aurel300 commented on prusti-encoder/src/encoders/mir_impure.rs:
```diff
@@ -764,22 +766,42 @@ impl<'vir, 'enc> mir::visit::Visitor<'vir> for EncoderVisitor<'vir, 'enc> {
+                    let pure_func_app = pure_func.function_ref.apply(self.vcx, &func_args);
+
+                    self.stmt(pure_func.return_type.method_assign.apply(self.vcx, [dest, pure_func_app]));
+                }
+                else {
```
```suggestion
                } else {
```

Aurel300 commented on prusti-encoder/src/encoders/mir_pure.rs:
```diff
@@ -18,6 +19,7 @@ pub enum MirPureEncoderError {
     UnsupportedTerminator,
 }
 
+// TODO: does this need to be `&'vir [..]`?
```
Is it the lifetime that's the problem? We could also add a `for` lifetime to the lazy methods, then use it as needed, e.g. here.

JonasAlaif replied:
I'd want a `&'a [vir::Expr<'vir>]` instead: the vector probably doesn't need to be interned. A `for` for the lazy methods doesn't work I'm pretty sure since you do need to get `&'vir VirCtxt` if you want to allocate stuff with a `'vir` lifetime

Aurel300 replied:
I meant to have both the `'vir` lifetime *and* another `for` lifetime for calls.

Aurel300 commented on prusti-encoder/src/encoders/mir_pure.rs:
```diff
@@ -445,39 +455,77 @@ impl<'tcx, 'vir: 'enc, 'enc> Encoder<'tcx, 'vir, 'enc>
                 enum PrustiBuiltin {
                     Forall,
+                    SnapshotEquality,
+                    Pure(DefId),
```
Rename the enum or make this (purity) a separate mechanism. Ideally the builtin stuff could be extracted and only deal with figuring out if a `DefId` is a Prusti builtin.

JonasAlaif replied:
I've moved it to a separate function for now (and removed the `Pure` variant), we can extract more later

Aurel300 commented on prusti-encoder/src/encoders/mir_impure.rs:
```diff
@@ -41,14 +41,18 @@ const ENCODE_REACH_BB: bool = false;
-    type TaskDescription<'vir> = DefId;
+    type TaskDescription<'vir> = (
+        DefId, // ID of the function
+        ty::GenericArgsRef<'vir>, // ? this should be the "signature", after applying the env/substs
+        Option<DefId>, // ID of the caller function, if any
```
Why does this matter for impure methods? Are we generating monomorphised methods (for now)?

JonasAlaif replied:
Yep

tillarnold commented:
Not sure if this is helpful or intended, Feel free to ignore but with my Vector code that previously worked 

```rust
extern crate prusti_contracts;
use prusti_contracts::*;

pub struct Vector {
    len: i32,
    contents: i32, // dummy variable encoding the contents of the vector
}

impl Vector {
    #[pure]
    #[trusted]
    #[requires(idx >= 0)]
    #[requires(idx < self.len)]
    #[requires(self.len >= 0)]
    fn get(self, idx: i32) -> i32 {
        unimplemented!()
    }

    #[trusted]
    #[requires(idx >= 0)]
    #[requires(idx < self.len)]
    #[requires(self.len >= 0)]
    #[ensures(self.len == result.len)]
    #[ensures(result.get(idx) == value)]
    #[ensures(forall(|i : i32| (i >= 0  && i < self.len && !(i == idx)) ==> result.get(i) == self.get(i)))]
    fn set(self, idx: i32, value: i32) -> Self {
        unimplemented!()
    }
}


fn main() {}

```

i now get an error
```
error: internal compiler error: compiler/rustc_middle/src/ty/mod.rs:2342:13: item_name: no name for DefPath { data: [DisambiguatedDefPathData { data: Impl, disambiguator: 0 }, DisambiguatedDefPathData { data: ValueNs("prusti_post_item_set_111f3974dc5e48b5b0480972e6810f5a"), disambiguator: 0 }, DisambiguatedDefPathData { data: ClosureExpr, disambiguator: 0 }], krate: crate0 }
```

JonasAlaif commented:
Hmm that's because the `get` is called from the forall closure, so we try to give the `get` function the following name `f_get_CALLER_{closure name here}` but the compiler doesn't have a name for the closure. The underlying issue is that we need a better system for naming things in Viper, but I'll make a temporary fix for now

JonasAlaif committed:
Update prusti-encoder/src/encoders/mir_impure.rs

Co-authored-by: Aurel <Aurel300@users.noreply.github.com>

JonasAlaif committed:
Fully differentiate `'tcx` and `'vir`

JonasAlaif committed:
Merge branch 'till_rebuilt_2' of https://github.com/tillarnold/prusti-dev into till_rebuilt_2

JonasAlaif committed:
Fix caller name issue

JonasAlaif committed:
Merge branch 'rewrite-2023' into till_rebuilt_2

JonasAlaif committed:
Resolve merge issues

JonasAlaif committed:
Add `base_spec` comment

JonasAlaif committed:
Extract Prusti builtin encoding

JonasAlaif committed:
More monomorphisation
